### ISSUE TITLE
gostatsd as a cluster-aware Kubernetes DaemonSet

### ISSUE THREAD

**Issue by mwhittington21** [atlassian/gostatsd#281]
I'd like to propose an additional module inside gostatsd which would be able to enrich packets with Kubernetes metadata based on which IP address sent a packet.

I'm looking to use gostatsd inside a Kubernetes cluster, one per node, to allow applications running on the node to send metrics to a locally running gostatsd which will be able to augment packets with tags including metadata about which application sent the metric before forwarding it on to an upstream aggregation layer.

This enrichment would come in much the same way as the cloud providers work inside gostatsd today. If inside Kubernetes and with a config file or flag enabling the feature, then incoming metrics would be enriched with data cached from the Kubernetes API server. Without getting too much into the specifics I see it working similar to the pod IP lookup cache inside something like [kube2iam](https://github.com/jtblin/kube2iam).

If this is a feature that you believe could fit into gostatsd then I'm willing to submit a PR making it happen.

**Comment by tiedotguy** [comment]
Sounds good to me. It **is** a cloud provider, really: a CP just an external map of ip -> tags.

My main suggestion would be to have a way to filter out tags (whitelist? prefix? you can also do regex because it's not high churn (expensive CPU)), as I know a lot of noise can be picked up via the EC2 tags.

That noise can be filtered out via filters in the pipeline, but it's better to do it once, rather than for every single metric.

**Comment by mwhittington21** [comment]
Agreed on the filtering. Here's a a few paragraphs that I came up with when I was brainstorming how this would work end to end.

---

Kubernetes pod annotation keys matching a certain prefix (e.g. `gostatsd.github.com/`) could have the second part of their annotation key used as a tag name for the gostatsd. The value of the annotation would then be used as the tag value.

So for example an annotation of `gostatsd.github.com/foo` with the value `bar` would have the gostatsd tag any metrics received from that pod with `foo:bar`.

We would need the following things to be customisable, either through command line flags or the config file:
- a whitelist of default tags to look for inside the annotations of a pod
  - optional - we can always self-discover from the prefix, but this may not be desireable as it can promote misuse
- a prefix for annotations which will be used to find the default tags
  - this would not be required if the whitelist is provided
- the location of a creds file to auth to k8s (if unsupplied then use the in-cluster auth via the mounted token)

The Kubernetes module inside gostatsd would handle setting up a WATCH on the API server, watching all pods running on the current node (or perhaps it would be better if this watch query was customisable for different use cases, like running one gostatsd for an entire cluster). It would then cache the tags for every pod running on the node against the IP address of those pods. When receiving a new statsd packet, it would then add the default tags if the sender IP address matches an entry in that dictionary. The packets would then be forwarded as normal.

**Comment by tiedotguy** [comment]
I've pretty much given up on command line flags - I never got them to work with viper, which this would plumb in to.

This whole approach seems pretty reasonable to me, although I'm not a k8s user, so it might be worth asking @dadux to sanity check it.

Would pod be the only level of data source? What about higher level (ie, node, cluster, data center, AZ, region, etc), or lower level (ie, container)?

I think it would be good if the watch could be customizable (I'm also planning this for clustering at some point, btw, but I was expecting to use labels), however I don't know how things would interact with a central service, rather than a daemon set. Would a more centralized service only have visibility of the node the packet came from, not the pod or container?

Just to be clear - are you planning this as a cloud provider module?

**Comment by dadux** [comment]
I pretty much agree with all what @mwhittington21 say and looking forward to have a k8s gostatsd 🎉 

However we should give the choice of having tags based on Kubernetes annotations OR labels. (or both)
Annotations probably works better for us, but most tools out there include all labels on their metrics.

**Comment by mwhittington21** [comment]
> Would pod be the only level of data source? What about higher level (ie, node, cluster, data center, AZ, region, etc), or lower level (ie, container)?

Pod and node would probably be the most useful things to include immediately as you can use the exact same interface to retrieve the metadata, just watching a different object. AZ/region can be determined by node labels so we could get those via the node watch.

Lower level things like containers don't have their own metadata so if we wanted to add tags from them it would be by whitelisting spec fields. I'll have a think about how to include this but probably won't implement it for the initial PR.

> I think it would be good if the watch could be customizable (I'm also planning this for clustering at some point, btw, but I was expecting to use labels), however I don't know how things would interact with a central service, rather than a daemon set. Would a more centralized service only have visibility of the node the packet came from, not the pod or container?

Totally agree on the customisable watch. I'll think about how best to expose this in the configuration file, probably a query based on labels.

In my mind the difference between a centralised service and a node-level service is just that it watches the whole cluster for objects, rather than just looking for objects related to the node it's running on. Otherwise it would work exactly the same way, having the same level of visibility.

> Just to be clear - are you planning this as a cloud provider module?

Yes, unless you have an idea for where it would better fit. I believe it fits the bill for one, although it might be beneficial for some use cases to be able to enable both the k8s and AWS cloud providers at once, and I'm not sure if that is possible currently.

**Comment by tiedotguy** [comment]
Groovy, that all around sounds good.

And yes, I agree it should be a cloud provider as well. At some far distant point I'm going to refactor the CP to use named providers, which for example would enable something like:
```
cloud-providers=['prod', 'dev']

[cloud-provider.prod]
type=aws
range=10.0.0.0/16
creds=metadata

[cloud-provider.dev]
type=aws
range=10.128.0.0/24
creds=jdgofd
```

Obviously needs a lot more thought, especially around what exactly it's trying to solve, but that's the rough idea, which also enables multiple providers in parallel.  Currently we can only get creds via the metadata API as well, which is not the safest for a k8s environment.

**Comment by mwhittington21** [comment]
That design for cloud providers looks great. Cloud provider it is!

In Kubernetes you could just give a path to an auth token in a file, which could be mounted via a Secret. I think the creds for Kube should be fairly straightforward.

**Comment by ash2k** [comment]
I think it all makes sense. I think there is some code in iam4kube that it would be beneficial to reuse to avoid polling informers and instead have blocking lookups from informers.
Please have a look at:
- https://github.com/ash2k/iam4kube/blob/master/pkg/kube/blocking_index.go - can be reused as is

- https://github.com/ash2k/iam4kube/blob/master/pkg/kube/kroler.go - usage example of the above. Note how we just block on informer lookups without any polling. Lookups are also cancellable.

I'm more than happy to help to build this functionality. I've built the CP code in gostatsd and this code above ^ so I can help integrate it all together.

**Comment by tiedotguy** [comment]
Just something that occurred to me - k8s is push, and CP is pull with a negative cache. Is there any chance of a race condition where the WATCH hasn't received and processed the data before the application sends the metric?

**Comment by mwhittington21** [comment]
Yes - in that case you would end up with the first few metrics sent lacking tags (or you could block for a small period to see if it resolves). We've seen this problem before in other things using WATCH caches. I can't remember if @ash2k's model handles it better than the more naive model that kube2iam uses.

**Comment by tiedotguy** [comment]
Ok, blocking would be the ideal method, but even then we need a timeout eventually.  There **may** also be some value in dropping rather than keeping without tags, but I'm not sure, and the CP can't do that at present anyway.

If it turns out that we do need to trigger the negative cache for a while, I would suggest exposing a way in the CP framework to change the negative TTL (unfortunately it's loaded from viper too early for you to override it like in [NewRelic](https://github.com/atlassian/gostatsd/blob/master/pkg/backends/newrelic/newrelic.go#L433)), thus allowing it to be minimized (since there's only one CP, you won't hurt anyone else, and this can be fixed when multi-CP comes it)

**Comment by ash2k** [comment]
Yes, the blocking index code was built to solve this exact problem :) It handles races and supports timeouts by design.

**Comment by ash2k** [comment]
So, I've started going down the `CloudProvider` route but then I realized - `CloudHandler` (the thing that is using `CloudProvider` to enrich metrics/events with tags) is doing all these things for us:

- 10ms batching to make fewer requests to the API
- rate limiting to avoid starving off other API users
- caching and negative caching of tags for an IP

But we don't need them! We'll have a `Pod` informer and it does the job even better.
So I was thinking about this and decided to try to build another implementation of the `PipelineHandler` interface. i.e. just a step in the metrics processing pipeline that enriches metrics/events with tags from the informer. I haven't finished coding, but it looks promising.
WDYT?

**Comment by tiedotguy** [comment]
Personally I think it should be in CP, partially for the organizational structure, but also CP should (but currently can't) be able to do things such as have a static list of tags it whitelists / blacklists (a filter can't do the whitelist aspect, plus it's better to never add them in the first place) which are common to all the "load from external" components.

If there's features which not all CPs need, then that's fine. If there's features which should be tuned for a specific CP, that can be done (like [NewRelic](https://github.com/atlassian/gostatsd/blob/master/pkg/backends/newrelic/newrelic.go#L433) overrides some defaults).

These are mostly excuses though...  I think for me comes down to, if I was explaining the architecture to someone, and told them about the pipelines and the cloud provider, and then told them about the k8s pipeline specifically, I'd expect to be asked "why isn't that a cloud provider?", but if it was the other way, I wouldn't expect "why isn't that a pipeline handler?"

Maybe not the best reason to say it should be a CP, but that's my logic :)

**Comment by mwhittington21** [comment]
I think Mikhail has the cleanest approach overall in terms of pure software engineering, but I'm more concerned about timeliness and fitting in with the existing codebase with these changes.

My current train of thought is to use the `Pod` informer as a cache to inform the generic cache provided by the `CloudHandler`. The `CloudHandler` handles the already processed tags, in a list of tags exactly like it retrieves from EC2 instances with the AWS handler.

This requires touching the least amount of code, and since it's only storing duplicate tags there isn't a lot of wastage. We can then tune the `CloudHandler` to access the cache in whatever volume makes sense for Kubernetes and it will automatically forward that through to the cache which will respond with the data as it comes in.

In my current implementation I've set a timeout for those cache accesses so if data isn't there then we don't waste too long waiting and just forward the packets, but we can modify this depending on user needs or even expose this as a setting.

Thoughts on this approach?

**Comment by ash2k** [comment]
I was thinking of the "common things that enrich metrics/events". I think we just need to refactor our abstractions a bit. I think today's `CloudHandler` can be replaced with "metrics enricher" which uses an interface to obtain tags/etc and there are two implementations of that interface:
- What was `CloudHandler`, refactored to fit the new API. Still uses `CloudProvider` under the hood
- Kubernetes-based impl.

That way we avoid the waste and get the benefits.

I wouldn't mind building what we have originally planned and then looking to refactor things to be more performant/less wasteful. If time is critical, then whatever delivers a solution first (but can be refactored later without breaking users!) is good enough.

**Comment by tiedotguy** [comment]
For speed of impl, I would assume @ash2k has the fastest approach. Certain refactorings of the CP are probably needed to get the caches to behave as expected, and deal with data freshness issues, and that whole thing can be bypassed by skipping the CP entirely.

If we're planning on making changes after, just remember to call out in the docs that we expect configuration options to change (I expect to change them anyway, with my plans I mentioned a few comments above)

**Comment by ash2k** [comment]
@mwhittington21 do you want to pair on finishing up your CP implementation? (I'm assuming it's in good shape, unlike mine) This will unblock your team and then we can refactor later.

**Comment by mwhittington21** [comment]
It's not really in great shape yet tbh, I spent a lot of time trying to fully understand the informers and your code from iam4kube and integrating just the cache into gostatsd. I'll message you on slack to organise further.

**Comment by mwhittington21** [comment]
I've submitted an implementation of this in the PR above, minus docs. I'll finalise the docs once the general approach has been discussed.

**Comment by ash2k** [comment]
Shall we close this issue now when the PR has been merged? Any outstanding work left to address it specifically? I think the refactoring I'm doing is not part of this issue - opened #290.

**Comment by mwhittington21** [comment]
Sounds good to me.
